app/main.py

```diff
@@ -0,0 +1,380 @@
+async def create_pipeline_form(pipeline_type: str, request: Request):
+    """Controller to render the create new pipeline form of a given type."""
+    params = config["pipelines"][pipeline_type]["params"]
+    context = {"request": request,
+               "params": params,
+               "pipeline_type": pipeline_type
+               }
+    return templates.TemplateResponse("new_pipeline.html", context)
+
+
+@app.post("/create_pipeline/", response_class=HTMLResponse)
+async def add_pipeline_from_form(request: Request,
+                                 db: Session = Depends(get_db),
+                                 ):
+    """Controller to process the response from create new pipeline form."""
+    form_data = await request.form()
+    form_data = jsonable_encoder(form_data)
+    prereq_pipe = form_data.pop("prereq_pipe")
+    prereq_step = form_data.pop("prereq_step")
+    pipeline = schemas.PipelineCreate(
+        template=form_data.pop("template"),
+        comment=form_data.pop("comment"),
+        prereq_pipe=int(prereq_pipe) if prereq_pipe != "" else None,
+        prereq_step=int(prereq_step) if prereq_step != "" else None,
+        params=form_data,
+    )
+    db_pipeline = crud.create_pipeline(db, pipeline)
+    context = {"request": request, "pipeline": db_pipeline}
+    return templates.TemplateResponse("view_pipeline.html", context)
+
+
+@app.get("/edit_pipeline/{pipeline_id}", response_class=HTMLResponse)
+async def edit_pipeline(pipeline_id: int,
+                        request: Request,
+                        db: Session = Depends(get_db)
+                        ):
+    """Controller to render the edit pipeline form."""
+    db_pipeline = crud.get_pipeline_by_id(db, pipeline_id)
+    if not db_pipeline:
+        raise HTTPException(
+            status_code=404,
+            detail=f"Pipeline with {pipeline_id=} does not exist."
+        )
+    context = {"request": request,
+               "pipeline": db_pipeline,
+               "status_options": models.PIPELINE_STATUSES,
+               "templates": list(config["pipelines"].keys()),
+               }
+    return templates.TemplateResponse("edit_pipeline.html", context)
+
+
+@app.post("/update_pipeline/", response_class=HTMLResponse)
+async def update_pipeline_from_form(request: Request,
+                                    db: Session = Depends(get_db),
+                                    ):
+    """Controller to process the response from the edit pipeline form."""
+    form_data = await request.form()
+    form_data = jsonable_encoder(form_data)
+
+    # Convert nested Dicts and replace empty strings with Nones:
+    if form_data["params"]:
+        form_data["params"] = loads(form_data["params"])
+    else:
+        form_data["params"] = None
+    if form_data["steps"]:
+        form_data["steps"] = loads(form_data["steps"])
+    else:
+        form_data["steps"] = None
+    if not form_data["prereq_pipe"]:

```

Comment:
`setdefault`

Reply:
I am not dealing with missing keys here. I am replacing existing, empty string values with None values, so those fields will be left unchanged when updating the record instead of being replaced with emtpy strings.

So I don't think setdefault would help me with that, would it?